bin/inx-chronicle/src/listener.rs

```diff
@@ -1,75 +1,198 @@
 // Copyright 2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-//! The [`InxListener`] subscribes to events from INX and forwards them as an [`InxEvent`] via a Tokio unbounded
+//! The [`InxListener`] subscribes to events from INX and forwards them via a Tokio unbounded
 //! channel.
 
-use futures::StreamExt;
+use std::{fmt::Debug, marker::PhantomData};
+
+use async_trait::async_trait;
+use chronicle::{
+    inx::{InxConfig, InxError},
+    runtime::{
+        actor::{context::ActorContext, envelope::HandleEvent, error::ActorError, handle::Addr, report::Report, Actor},
+        config::ConfigureActor,
+        error::RuntimeError,
+    },
+};
 use inx::{
     client::InxClient,
     proto::{MessageFilter, NoParams},
-    Channel,
+    Channel, Status,
 };
 use log::info;
-use tokio::task::JoinHandle;
+use thiserror::Error;
 
-use crate::broker::BrokerAddr;
+use crate::broker::Broker;
 
-#[derive(Debug)]
-pub enum InxEvent {
-    Message(inx::proto::Message),
-    LatestMilestone(inx::proto::Milestone),
+#[derive(Debug, Error)]
+pub enum InxListenerError {
+    #[error("The broker actor is not running")]
+    MissingBroker,
+    #[error(transparent)]
+    Inx(#[from] InxError),
+    #[error(transparent)]
+    Read(#[from] Status),
+    #[error(transparent)]
+    Runtime(#[from] RuntimeError),
 }
 
+#[derive(Debug)]
 pub struct InxListener {
-    // TODO: Consider storing `inx_client` to potentially restart some streams.
-    message_handle: JoinHandle<()>,
-    latest_milestone_handle: JoinHandle<()>,
+    config: InxConfig,
+    broker_handle: Addr<Broker>,
 }
 
 impl InxListener {
-    pub fn new(inx_client: InxClient<Channel>, broker_addr: BrokerAddr) -> Self {
-        let message_handle = {
-            let mut inx_client = inx_client.clone();
-            let broker_addr = broker_addr.clone();
-            tokio::spawn(async move { subscribe_messages(&mut inx_client, broker_addr).await })
-        };
-
-        let latest_milestone_handle = {
-            let mut inx_client = inx_client;
-            tokio::spawn(async move { subscribe_latest_milestone(&mut inx_client, broker_addr).await })
-        };
-
-        Self {
-            message_handle,
-            latest_milestone_handle,
+    pub fn new(config: InxConfig, broker_handle: Addr<Broker>) -> Self {
+        Self { config, broker_handle }
+    }
+}
+
+#[async_trait]
+impl Actor for InxListener {
+    type Data = InxClient<Channel>;
+    type Error = InxListenerError;
+
+    async fn init(&mut self, cx: &mut ActorContext<Self>) -> Result<Self::Data, Self::Error> {
+        info!("Connecting to INX...");
+        let mut inx_client = self.config.build().await?;
```

Comment:
I still think something like `InxClient<Channel>::new(&self.config).await?` would be more future-proof.

Reply:
Take it up with @grtlr